enigma-tools-u/src/web3_utils/contract_ext.rs

```diff
@@ -0,0 +1,146 @@
+/// be `?`-cast into `failure::Fail`
+#[derive(Debug, failure::Fail)]
+pub enum Error {
+    #[fail(display = "{}", _0)]
+    Web3Error(String),
+
+    #[fail(display = "{}", _0)]
+    Web3ContractError(String),
+
+    #[fail(display = "{}", _0)]
+    EthabiError(String),
+
+    #[fail(display = "{}", _0)]
+    IoError(#[cause] std::io::Error),
+
+    #[fail(display = "{}", _0)]
+    SerdeJsonError(#[cause] serde_json::Error),
+
+    #[fail(display = "{}", _0)]
+    ParseIntError(#[cause] std::num::ParseIntError),
+}
+
+macro_rules! from_error_to_string {
+    ($variant: path, $source: ty) => {
+        impl From<$source> for Error {
+            fn from(other: $source) -> Self {
+                $variant(other.to_string())
+            }
+        }
+    }
+}
+
+from_error_to_string!(Error::Web3Error, web3::Error);
+from_error_to_string!(Error::Web3ContractError, web3::contract::Error);
+from_error_to_string!(Error::EthabiError, ethabi::Error);
+
+macro_rules! from_error {
+    ($variant: path, $source: ty) => {
+        impl From<$source> for Error {
+            fn from(other: $source) -> Self {
+                $variant(other)
+            }
+        }
+    }
+}
+
+from_error!(Error::IoError, std::io::Error);
+from_error!(Error::SerdeJsonError, serde_json::Error);
+from_error!(Error::ParseIntError, std::num::ParseIntError);
+
+fn option_to_u256(value: Option<U256>) -> ethereum_types::U256 {
+    value.unwrap_or_else(|| ethereum_types::U256::from(0))
+}
+
+/// Confused? You should be. Please read the documentation in the top of this file.
+pub fn signed_call_with_confirmations<T: Transport, P>(
+    // general tools
+    web3: &web3::api::Web3<T>,
+    // contract details
+    contract_abi: &ethabi::Contract,
+    contract_address: Address,
+    // sender details
+    private_key: H256,
+    from: Address, // Should belong to the private key
+    // function details
+    func: &str,
+    params: P,
+    // network details
+    options: Options,
+    chain_id: u8,
+    confirmations: usize,
+) -> Result<confirm::SendTransactionWithConfirmation<T>, Error>
+    where
+        P: Tokenize,
+{
+    let poll_interval = std::time::Duration::from_secs(1);
+
+    let function = contract_abi.function(func)?;
+    let fn_data = function.encode_input(&params.into_tokens())?;
+
+    let nonce = if let Some(nonce) = options.nonce {
+        nonce
+    } else {
+        web3.eth().transaction_count(from, Some(BlockNumber::Latest)).wait()?
+    };
+
+    let tx = ethereum_tx_sign::RawTransaction {
+        nonce: nonce,
+        to: Some(contract_address),
+        value: option_to_u256(options.value),
+        gas_price: option_to_u256(options.gas_price),
+        gas: option_to_u256(options.gas),
+        data: fn_data,
+    };
+
+    let signed_tx = tx.sign(&private_key, &chain_id);
+
+    Ok(confirm::send_raw_transaction_with_confirmation(
+        web3.eth().transport().clone(),
+        Bytes::from(signed_tx),
+        poll_interval,
```

Comment:
what is the purpose of this interval and why is it of 1 sec?

Reply:
This is the same interval used in the original implementation [here](https://github.com/tomusdrw/rust-web3/blob/c69bf938a0d3cfb5b64fca5974829408460e6685/src/contract/mod.rs#L153)